### ISSUE TITLE
compute-vm: Secure tags with "For use with network firewall"  can't be used and creates crash

### ISSUE THREAD

**Issue by hoppefamily** [GoogleCloudPlatform/cloud-foundation-fabric#1887]
**Describe the bug**
If a secure tag, where "For use with network firewall" has been set, is used in variable tag_bindings it creates  a crash.

**Result**
Error: Error waiting for instance to create: Internal error. Please try again or contact Google Support. (Code: 'Internal error')

**Comment by hoppefamily** [comment]
Tags of that kind can not be added to disks.

**Comment by juliocc** [comment]
This sounds more like a provider issue.

Can you share your code? Please provide all the details requested when you open a bug.

**Comment by hoppefamily** [comment]
I don't think the code itself would be any helpful. Let me instead explain how to replicate the error:
- Create a secure tag via console and set the "For use with network firewall" flag.
- Create a VM via the compute-vm terraform module and set the tag_bindings variable like:
tag_bindings = {
  "tagKeys/281483402711742" = "tagValues/281479473111820"
}
- deploy the VM. It will crash with:
Error: Error waiting for instance to create: Internal error. Please try again or contact Google Support. (Code: 'Internal error')

On the one side it is a provider or even API issue.
On the other side (and that is the reason I opened here this issue) it is obvious that a secure tag which used in firewall rules makes no sense to e attached to a disk. I saw it also stated somewhere in the documentation but can not find the place anymore.
But it makes perfect sense to attach a tag which is use in firewall rules to an instance. The modules in the current form does not allow to attach these kind of tag solely to the instance and not to the boot disk. The module always attaches the tags to the instance AND the boot disk. This is the issue.

**Comment by juliocc** [comment]
From your description it seems the issue comes from [this line](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/modules/compute-vm/main.tf#L222)

@ludoo wdyt if we keep the existing `tag_bindings` only for instance-level tags? If needed we can later introduce another variable (or field) for disk-level tags.

**Comment by wiktorn** [comment]
I think this might be a short example for that:
```hcl
module "org" {
  source          = "./fabric/modules/organization"
  organization_id = var.organization_id
  tags = {
    environment = {
      description = "Environment specification."
      iam         = null
      values = {
        dev  = null
        prod = null
      }
    }
  }
}

module "simple-vm-example" {
  source     = "./fabric/modules/compute-vm"
  project_id = var.project_id
  zone       = "${var.region}-b"
  name       = "test"
  network_interfaces = [{
    network    = var.vpc.self_link
    subnetwork = var.subnet.self_link
  }]
  tag_bindings = {
    (module.org.tag_keys["environment"].id) = module.org.tag_values["environment/prod"].id
  }
}

```
I tried this one, and actually no failures occured.

**Comment by ludoo** [comment]
> From your description it seems the issue comes from [this line](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/modules/compute-vm/main.tf#L222)
> 
> @ludoo wdyt if we keep the existing `tag_bindings` only for instance-level tags? If needed we can later introduce another variable (or field) for disk-level tags.

yep it makes sense, I also remember discussing some weirdness and/or incomplete support with instances and tags internally not too long ago, maybe we should resurface that discussion and attach it to a bug here

**Comment by ludoo** [comment]
Scrap my comment, disks and instances need the same tag-level support or IAM conditions won't work.

What Wiktor mentions above is what I remember: tag-based support is for resource management tags, not VPC-scoped tags.

**Comment by wiktorn** [comment]
So, do I understand correctly, that we have two separate use cases which use Resource Manager tags here:
1. For IAM conditions, which requires the tags to be set on instance and disk level (preferably the same)
2. For firewall policies, which requires VPC-scoped tags assigned only on instance level (that's how I understood this [docs](https://cloud.google.com/vpc/docs/create-manage-tags-vpc-resources))

So to support both of the use cases we need to allow to set the tags only on instance level, and explain to the users, which to use when, or have a separate field `vpc_scoped_tag_bindings` (not pretty) which goes only to the instance level.

**Comment by juliocc** [comment]
> Yes. But for network scoped tags there's some complications in the provider, or there were up to a few weeks ago: from my recollection you can set them in the instance resource, but then any change triggers recreation. While the external binding resource is not available, or did not work last time I tried it. I need to fetch the chat history for the discussion.

This is still true and is [documented here](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_instance#resource_manager_tags:~:text=a%20specific%20tag.-,This%20value%20is%20not%20returned%20by%20the%20API.%20In%20Terraform%2C%20this%20value%20cannot%20be%20updated%20and%20changing%20it%20will%20recreate%20the%20resource.,-The%20scratch_disk%20block)

**Comment by hoppefamily** [comment]
> From your description it seems the issue comes from [this line](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/modules/compute-vm/main.tf#L222)
> 
> @ludoo wdyt if we keep the existing `tag_bindings` only for instance-level tags? If needed we can later introduce another variable (or field) for disk-level tags.

yes

**Comment by ludoo** [comment]
Nope. 😀 The current interface is used across modules for resource management scoped tags. We need to define a new one for network scoped tags.

**Comment by drebes** [comment]
`network_tag_bindings`?

**Comment by drebes** [comment]
> `network_tag_bindings`?

Or actually `firewall_tag_bindings` to not confuse with the old school network tags.
